### PR TITLE
feat(web-analytics): Turn timezones into UTC offsets instead of timezone names

### PR DESCRIPTION
**Pull request by rafaeelaudibert**
Timezone names on their own aren't that useful because they're basically a harder-to-use version of displaying events by city/region. Displaying them by offsets, however, can prove to be valuable. We're doing some magic to calculate the offset instead of simply using something like `timezoneOffset` because `toString` is one of the only functions in Clickhouse that allows us to use a dynamic timezone versus a hardcoded one. This might have some performance implications, we'll figure it out soon :). To workaround that, we're converting to a string, then parsing that back to a date, and getting the unix timestamp. Comparing the UNIX timestamp in a specific timezone versus the value at UTC allows us to get the offset at that day.

See #26376

## Changes

image

## Does this work well for both Cloud and self-hosted?

Yep

## How did you test this code?

Comprehensive unit test + manual UI testing

### CODE REVIEW

File: posthog/hogql_queries/web_analytics/test/test_web_stats_table.py [hunk 1]
```diff
@@ -1006,50 +1006,67 @@ def test_cohort_test_filters(self):
...
+            for i in range(idx + 1):
+                _create_event(
+                    team=self.team,
+                    event="$pageview",
+                    distinct_id=distinct_id,
+                    timestamp=date,
+                    properties={"$session_id": session_id, "$pathname": f"/path{i}", "$timezone": distinct_id},
+                )
+
+        with self.assertRaisesRegex(Exception, "Cannot load time zone"):
```
Comment (rafaeelaudibert) [inline posthog/hogql_queries/web_analytics/test/test_web_stats_table.py:1067]: Reasonable error?

File: posthog/hogql_queries/web_analytics/test/test_web_stats_table.py [hunk 2]
```diff
@@ -1006,50 +1006,67 @@ def test_cohort_test_filters(self):
         assert results == [["/path1", 1, 1]]
 
     def test_timezone_filter(self):
-        d1, s1 = "d1", str(uuid7("2024-07-30"))
-        d2, s2 = "d2", str(uuid7("2024-07-30"))
+        date = "2024-07-30"
```
Comment (robbie-c) [inline posthog/hogql_queries/web_analytics/test/test_web_stats_table.py:1009]: For me to be *suuuper* confidant this works, I'd like to see some tests that cover before / after the daylight savings change for the same timezone

File: posthog/hogql_queries/web_analytics/stats_table.py [hunk 3]
```diff
@@ -77,6 +77,7 @@ def to_main_query(self) -> ast.SelectQuery:
 )
 GROUP BY "context.columns.breakdown_value"
 ORDER BY "context.columns.visitors" DESC,
+"context.columns.views" DESC,
```
Comment (rafaeelaudibert) [inline posthog/hogql_queries/web_analytics/stats_table.py:80]: @robbie-c Do you see any problem with this new clause here? It's helpful when different fields have the same number of visitors, but a different number of views. It's more helpful during tests than on production data. Anything against the change?

File: posthog/hogql_queries/web_analytics/test/test_web_stats_table.py (hunk 2 above)
Comment (rafaeelaudibert) [reply posthog/hogql_queries/web_analytics/test/test_web_stats_table.py:1009]: Just added it, good call

File: posthog/hogql_queries/web_analytics/stats_table.py [hunk 4]
```diff
@@ -500,9 +505,11 @@ def _counts_breakdown_value(self):
             case WebStatsBreakdown.CITY:
                 return parse_expr("tuple(properties.$geoip_country_code, properties.$geoip_city_name)")
             case WebStatsBreakdown.TIMEZONE:
-                # Timezone offsets would be slightly more useful, but that's not easily achievable
-                # with Clickhouse, we might attempt to change this in the future
-                return ast.Field(chain=["properties", "$timezone"])
+                # Get the difference between the UNIX timestamp at UTC and the UNIX timestamp at the event's timezone
+                # Value is in milliseconds, turn it to hours, works even for fractional timezone offsets (I'm looking at you, Australia)
+                return parse_expr(
+                    "if(isNull(properties.$timezone), NULL, (toUnixTimestamp64Milli(parseDateTimeBestEffort(assumeNotNull(toString(timestamp, properties.$timezone)))) - toUnixTimestamp64Milli(timestamp)) / 3600000)"
```
Comment (robbie-c) [inline posthog/hogql_queries/web_analytics/stats_table.py:511]: Let's add a test for key not present, value is NULL, and value is an empty string, I think you might need some extra logic to handle the empty string.

**Review by robbie-c — APPROVED** [review]
LGTM, just one more test IMO but happy for this to be a follow-up :)

**Pull request merged**